File: src/rex/states/confirming.py
"""
Awaiting confirmation state handler.

Handles human-in-the-loop confirmation for tool calls that
require explicit user approval (e.g., creating reminders).
"""

from typing import TYPE_CHECKING

from agent import PendingConfirmation, confirm_tool_call
from core.state_machine import ConversationState, StateHandler, StateResult

from .phrases import is_confirmation

if TYPE_CHECKING:
    from core.context import AppContext
    from stt import Transcriber
    from tts import InterruptibleSpeaker
    from wake_word import WakeWordListener

# Timeout for confirmation responses
CONFIRMATION_TIMEOUT = 10.0


class AwaitingConfirmationHandler(StateHandler):
    """
    Handler for the AWAITING_CONFIRMATION state.

    Prompts the user to confirm or reject a pending tool call,
    then processes the response.
    """

    def __init__(
        self,
        listener: "WakeWordListener",
        transcriber: "Transcriber",
        speaker: "InterruptibleSpeaker",
    ):
        self._listener = listener
        self._transcriber = transcriber
        self._speaker = speaker
        self._pending: PendingConfirmation | None = None

    @property
    def state(self) -> ConversationState:
        return ConversationState.AWAITING_CONFIRMATION

    def enter(self, ctx: "AppContext", data: dict | None = None) -> None:
        """Store the pending confirmation."""
        if data:
            self._pending = data.get("pending")
        else:
            self._pending = None

    def process(self, ctx: "AppContext") -> StateResult:
        """
        Handle the confirmation flow.

        Returns:
            SPEAKING with the result of confirmation/rejection
        """
        if self._pending is None:
            return StateResult(
                next_state=ConversationState.SPEAKING,
                data={"response": "Something went wrong with the confirmation."},
            )

        # Speak the confirmation prompt
        print(f"\n🤖 Rex: {self._pending.confirmation_prompt}\n")
        self._speaker.speak_interruptibly(self._pending.confirmation_prompt)

        # Listen for response
        print("🎤 Listening for confirmation...")
        audio = self._listener.listen_for_speech(timeout=CONFIRMATION_TIMEOUT)

        if audio is None:
            # No response - treat as rejection
            print("⏱️ No confirmation received, cancelling.")
            response, history = confirm_tool_call(self._pending, confirmed=False)
            ctx.conversation_history = history
            return StateResult(
                next_state=ConversationState.SPEAKING,
                data={"response": response, "force_end_conversation": True},
            )

        transcription = self._transcriber.transcribe(audio)
        if not transcription:
            print("⏱️ Could not understand response, cancelling.")
            response, history = confirm_tool_call(self._pending, confirmed=False)
            ctx.conversation_history = history
            return StateResult(
                next_state=ConversationState.SPEAKING,
                data={"response": response, "force_end_conversation": True},
            )

        print(f"\n💬 You said: {transcription}\n")

        if is_confirmation(transcription):
            print("✅ Confirmed!")
            response, history = confirm_tool_call(self._pending, confirmed=True)
        else:
            print("❌ Cancelled.")
            response, history = confirm_tool_call(self._pending, confirmed=False)

        ctx.conversation_history = history

        return StateResult(
            next_state=ConversationState.SPEAKING,
            data={"response": response, "force_end_conversation": True},
        )

    def exit(self, ctx: "AppContext") -> None:
        """Clear pending confirmation."""
        self._pending = None

File: internal/states/confirming.go
package states

import (
	"context"
	"log/slog"
	"time"
)

const confirmationTimeout = 10 * time.Second

// ConfirmingHandler implements the AWAITING_CONFIRMATION state.
// It prompts the user to approve or reject a pending tool call and
// forwards the decision to the agent.
type ConfirmingHandler struct {
	deps    *Deps
	pending *PendingConfirmation
}

// NewConfirmingHandler creates a ConfirmingHandler.
func NewConfirmingHandler(deps *Deps) *ConfirmingHandler {
	return &ConfirmingHandler{deps: deps}
}

func (h *ConfirmingHandler) State() int { return StateAwaitingConfirmation }

// Enter stores the pending confirmation passed from the processing state.
func (h *ConfirmingHandler) Enter(data map[string]any) {
	h.pending = nil
	if data == nil {
		return
	}
	if v, ok := data["pending"].(*PendingConfirmation); ok {
		h.pending = v
	}
}

// Process speaks the confirmation prompt, captures the user's response,
// and relays the decision to the agent.
func (h *ConfirmingHandler) Process(ctx context.Context) StateResult {
	if h.pending == nil {
		return StateResult{
			NextState: StateSpeaking,
			Data: map[string]any{
				"response": "Something went wrong with the confirmation.",
			},
		}
	}

	slog.Info("asking for confirmation", "prompt", h.pending.ConfirmationPrompt)

	transcription := speakAndCapture(ctx, h.deps, h.pending.ConfirmationPrompt, confirmationTimeout)
	if transcription == "" {
		slog.Info("no confirmation received, cancelling")
		return h.reject(ctx)
	}

	slog.Info("confirmation response", "text", transcription)

	confirmed := IsConfirmation(transcription)
	if confirmed {
		slog.Info("tool call confirmed")
	} else {
		slog.Info("tool call rejected")
	}

	result, err := h.deps.Agent.ConfirmToolCall(ctx, h.pending, confirmed, h.deps.History)
	if err != nil {
		slog.Error("agent confirm error", "err", err)
		return StateResult{
			NextState: StateSpeaking,
			Data: map[string]any{
				"response":               "Sorry, I had trouble processing the confirmation.",
				"force_end_conversation": true,
			},
		}
	}

	h.deps.History = result.History
	return StateResult{
		NextState: StateSpeaking,
		Data: map[string]any{
			"response":               result.Response,
			"force_end_conversation": true,
		},
	}
}

// reject sends a rejection to the agent and returns a speaking result.
func (h *ConfirmingHandler) reject(ctx context.Context) StateResult {
	result, err := h.deps.Agent.ConfirmToolCall(ctx, h.pending, false, h.deps.History)
	if err != nil {
		slog.Error("agent reject error", "err", err)
		return StateResult{
			NextState: StateSpeaking,
			Data: map[string]any{
				"response":               "Cancelled.",
				"force_end_conversation": true,
			},
		}
	}
	h.deps.History = result.History
	return StateResult{
		NextState: StateSpeaking,
		Data: map[string]any{
			"response":               result.Response,
			"force_end_conversation": true,
		},
	}
}

// Exit clears the pending confirmation.
func (h *ConfirmingHandler) Exit() {
	h.pending = nil
}
